### PR TITLE
[V2.0] error handling WIP

### PR DESCRIPTION
**Pull request by kwek20**
# Description of change

## Type of change

- Enhancement (a non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Change checklist

Add an `x` to the boxes that are relevant to your changes, and delete any items that are not.

- [] I have followed the contribution guidelines for this project
- [] I have performed a self-review of my own code
- [] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation
- [] I have added tests that prove my fix is effective or that my feature works
- [] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: lets/src/transport/tangle.rs [hunk 1]
```diff
@@ -72,19 +78,28 @@ where
             .with_index(address.to_msg_index())
             .with_data(msg.into())
             .finish()
-            .await?
+            .await
+            .map_err(|e| Error::IotaClient("sending message", e))?
```
Comment (DyrellC) [inline lets/src/transport/tangle.rs:82]: This and the one below are a little inconsistent. Should probably stick to either 
`"sending message" and "receiving messages"` or `"send_message" and "recv_message"` with the appropriate change to the `Error::IotaClient` message syntactically

File: spongos/src/error.rs [hunk 2]
```diff
@@ -1,39 +1,65 @@
...
     //////////
-    /// Size of vec/array does not match (expected: {0}, found: {1})
+    #[error("Size of vec/array does not match (expected: {0}, found: {1})")]
     LengthMismatch(usize, usize),
 
     //////////
     // DDML Wrap/Unwrap
     //////////
-    /// There was an issue with the calculated signature, cannot unwrap message
+    #[error("here was an issue with the calculated signature, cannot unwrap message")]
```
Comment (DyrellC) [inline spongos/src/error.rs:19]:
```suggestion
    #[error("There was an issue with the calculated signature, cannot unwrap message")]
```

File: spongos/src/error.rs [hunk 3]
```diff
@@ -1,39 +1,65 @@
...
     // DDML IO
     //////////
-    /// Not enough space allocated for output stream (expected: {0}, found: {1})
+    #[error("Not enough space allocated for output stream (expected: {0}, found: {1})")]
     StreamAllocationExceededOut(usize, usize),
-    /// Not enough space allocated for input stream (expected: {0}, found: {1})
+    #[error("Not enough space allocated for input stream (expected: {0}, found: {1})")]
     StreamAllocationExceededIn(usize, usize),
+
+    #[error("Generating context {0} failed due to {1}")]
```
Comment (DyrellC) [inline spongos/src/error.rs:48]: I think this is the most confusing error to understand as is. In most implementations it reads like this: 
`Generating context ContentVerify failed due to [error]` which doesn't really make any sense. We should probably rephrase it to something along the lines of 

 ```suggestion
    #[error("Context failed to perform the message command \"{0}\"; Error: {1}")]
```

This way it would read `Context failed to perform the message command "ContentVerify"; Error: [error]`

File: streams/src/error.rs [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+#[derive(Debug, Error)]
+/// Error type of the iota client crate.
+#[allow(clippy::large_enum_variant)]
+/// #[serde(tag = "type", content = "error")]
+pub enum Error {
+    //////////
+    // Streams
+    //////////
+    #[error(
+        "address already taken for. The address '{1}' where the {0} message is being sent already contains some data, possibly spam."
```
Comment (DyrellC) [inline streams/src/error.rs:31]:
```suggestion
        "Address already taken. The address '{1}' where the {0} message is being sent already contains some data, possibly spam."
```

File: streams/src/error.rs [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+    )]
+    AddressUsed(&'static str, Address),
+
+    #[error("unexpected message type {0}")]
```
Comment (DyrellC) [inline streams/src/error.rs:35]:
```suggestion
    #[error("Unexpected message type {0}")]
```

File: streams/src/error.rs [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+    MessageTypeUnknown(u8),
+
+    #[error("message '{0}' not found in {1}")]
```
Comment (DyrellC) [inline streams/src/error.rs:38]:
```suggestion
    #[error("Message '{0}' not found in {1}")]
```

File: streams/src/error.rs [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+    MessageMissing(MsgId, &'static str),
+
+    #[error("Failed getting messages for {0} due to {1}")]
```
Comment (DyrellC) [inline streams/src/error.rs:41]:
```suggestion
    #[error("Failed to get messages.  Error: {0}")]
```
I don't think we need to specify which function caused this error, it'll already be traced

File: streams/src/error.rs [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+    Messages(&'static str, anyhow::Error),
+
+    #[error("Missing {0} for {1} in order to {2}")]
+    MissingUserData(&'static str, &'static str, &'static str),
```
Comment (DyrellC) [inline streams/src/error.rs:45]: This is never used from what I've seen

File: streams/src/error.rs [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+
+    #[error(
+        "user does not have a cursor in branch '{0}'. This probably means the user does not have write permission over that branch"
```
Comment (DyrellC) [inline streams/src/error.rs:48]:
```suggestion
        "User does not have a cursor stored in branch '{0}'. This probably means the user does not have write permission within that branch"
```

File: streams/src/error.rs [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+    )]
+    NoCursor(Topic),
+
+    #[error("the user does not have an identity, but needs one to {0}")]
```
Comment (DyrellC) [inline streams/src/error.rs:52]:
```suggestion
    #[error("User does not have an identity, but needs one to {0}")]
```

File: streams/src/error.rs [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+    NoIdentity(&'static str),
+
+    #[error("User identity contained no secret key")]
```
Comment (DyrellC) [inline streams/src/error.rs:55]:
```suggestion
    #[error("User identity contains no secret key")]
```

File: streams/src/error.rs [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+    NoSecretKey,
+
+    #[error(
+        "not connected to a stream. A user must either create a stream or connect to an existing one before attempting to {0}"
```
Comment (DyrellC) [inline streams/src/error.rs:59]:
```suggestion
        "Not connected to a stream. A user must either create a stream or connect to an existing one before attempting to {0}"
```

File: streams/src/error.rs [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+    )]
+    NoStream(&'static str),
+
+    #[error(
+        "Message not linked. The {0} message at address '{1:#?}' is not linked to any message. \
```
Comment (DyrellC) [inline streams/src/error.rs:64]:
```suggestion
        "Message not linked. The {0} message at address '{1:#?}' is not linked to a previous message. \
```

File: streams/src/error.rs [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+Any {0} message must be linked to a previous message by including the address of the previous message in the header"
```
Comment (DyrellC) [inline streams/src/error.rs:65]:
```suggestion
Any {0} message must be linked to a previous message by including the address of the existing message in the header"
```

File: streams/src/error.rs [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+    )]
+    NotLinked(&'static str, Address),
+
+    #[error("A payload must be specified in order to send a message")]
+    PayloadEmpty,
+
+    #[error("Setup error: {0}")]
+    Setup(&'static str),
+
+    #[error("topic {0} not found in store")]
```
Comment (DyrellC) [inline streams/src/error.rs:75]:
```suggestion
    #[error("Topic {0} not found in store")]
```

File: streams/src/error.rs [hunk 16]
```diff
@@ -0,0 +1,122 @@
...
+    TopicNotFound(Topic),
+
+    #[error("Transport error whilest doing {0} for address {1}: {2}")]
```
Comment (DyrellC) [inline streams/src/error.rs:78]:
```suggestion
    #[error("Transport error while trying to {0} for address {1}; Error: {2}")]
```

**Review by DyrellC — CHANGES_REQUESTED** [review]
I think the whole DID wrapping enum can be removed by using the `ToString` approach for converting the identity errors. Aside from that it's mostly suggestions on potential improvements to readability and messaging.

File: spongos/src/error.rs (hunk 3 above)
Comment (kwek20) [reply spongos/src/error.rs:48]: Yea this one was a hard one, as it feels unneeded but needs to be thrown from the spongos error throughout streams/lets

File: lets/src/id/did/url_info.rs [hunk 17]
```diff
@@ -135,10 +140,13 @@ where
             .mask(Bytes::new(&mut exchange_fragment_bytes))?
             .mask(Bytes::new(&mut signing_fragment_bytes))?;
 
-        *url_info.did_mut() = String::from_utf8(did_bytes)?;
-        *url_info.client_url_mut() = String::from_utf8(client_url)?;
-        *url_info.exchange_fragment_mut() = String::from_utf8(exchange_fragment_bytes)?;
-        *url_info.signing_fragment_mut() = String::from_utf8(signing_fragment_bytes)?;
+        let utf =
```
[the remark this answers is not in the document]
Comment (kwek20) [reply lets/src/id/did/url_info.rs:143]: Its more specific now but maybe a bit over the top
